contracts/contracts/coordination/Coordinator.sol

```diff
@@ -267,4 +326,55 @@ contract Coordinator is Ownable {
     ) external view returns (Participant memory) {
         return getParticipantFromProvider(rituals[ritualID], provider);
     }
+
+    function processRitualPayment(uint256 ritualID, address[] calldata providers, uint32 duration) internal {
+        uint256 ritualCost = feeModel.getRitualInitiationCost(providers, duration);
+        if (ritualCost > 0){
+            totalPendingFees += ritualCost;
+            assert(pendingFees[ritualID] == 0);  // TODO: This is an invariant, not sure if actually needed
+            pendingFees[ritualID] += ritualCost;
+            IERC20 currency = IERC20(feeModel.currency());
+            currency.transferFrom(msg.sender, address(this), ritualCost);
+            // TODO: Define methods to manage these funds
+        }
+    }
+
+    function processPendingFee(uint256 ritualID) public {
+        Ritual storage ritual = rituals[ritualID];
+        RitualState state = getRitualState(ritual);
+        require(
+            state == RitualState.TIMEOUT ||
+            state == RitualState.INVALID ||
+            state == RitualState.FINALIZED,
+            "Ritual is not ended"
+        );
+        uint256 pending = pendingFees[ritualID];
+        require(pending > 0, "No pending fees for this ritual");
+
+        // Finalize fees for this ritual
+        totalPendingFees -= pending;
+        delete pendingFees[ritualID];
+        // Transfer fees back to initiator if failed
+        if(state == RitualState.TIMEOUT || state == RitualState.INVALID){
+            // Amount to refund depends on how much work nodes did for the ritual.
+            // TODO: Validate if this is enough to remove griefing attacks
+            uint256 executedTransactions = ritual.totalTranscripts + ritual.totalAggregations;
+            uint256 expectedTransactions = 2 * ritual.dkgSize;
+            uint256 consumedFee = pending * executedTransactions / expectedTransactions;
+            uint256 refundableFee = pending - consumedFee;
+            IERC20 currency = IERC20(feeModel.currency());
+            currency.transferFrom(address(this), ritual.initiator, refundableFee);
```

Comment:
Nitpick, but `safeTransferFrom` would be better

Reply:
✅